horovod/run/run.py

```diff
@@ -322,35 +323,61 @@ def run():
     tmout = timeout.Timeout(start_timeout)
 
     key = secret.make_secret_key()
+    remote_host_names = []
     if args.host:
         if args.verbose:
-            print("Checking ssh on all hosts.")
-        # Check if we can ssh into all hosts successfully. This is required for
-        _check_all_hosts_ssh_successful(host_addresses, args.ssh_port,
+            print("Rresolving remote host names.")
+        remote_host_names = network.filter_local_addresses(host_names)
+
+        if args.verbose:
+            print("Checking ssh on all remote hosts.")
+        # Check if we can ssh into all remote hosts successfully.
+        _check_all_hosts_ssh_successful(remote_host_names, args.ssh_port,
                                         fn_cache=fn_cache)
         if args.verbose:
-            print("SSH was successful into all the hosts.")
+            print("SSH was successful into all the remote hosts.")
 
+        hosts_arg = "-H {hosts}".format(hosts=args.host)
+    else:
+        # if user does not specify any hosts, mpirun by default uses local host.
+        # There is no need to specify localhost.
+        hosts_arg = ""
+
+    if args.host and len(remote_host_names) > 0:
         if args.verbose:
             print("Testing interfaces on all the hosts.")
-        # Find the set of common, routed interfaces on all the hosts and specify
-        # it to be used by NCCL.
-        common_intfs = _driver_fn(key, host_addresses, tmout,
+        # If there is at least one remote host name, find the set of common,
+        # routed interfaces on all the hosts and specify it in the args to be
+        # used by NCCL. It is expected that the following function will find
+        # at least one interface otherwise, it will raise an exception.
+        common_intfs = _driver_fn(key,
+                                  remote_host_names,
+                                  tmout,
                                   args.ssh_port,
                                   args.verbose,
                                   fn_cache=fn_cache)
-        if args.verbose:
-            print("Interfaces on all the hosts were successfully checked.")
 
-        hosts_arg = "-H {hosts}".format(hosts=args.host)
+        if len(remote_host_names) != len(host_names):
```

Comment:
Does this fork in logic mean that if we have two hosts, one of which is local host, we won't do circular ping where localhost tries to reach interfaces on host2?

Alternative idea - what if we keep existing logic, but the only difference is if we're trying to `ssh` to localhost, we just skip ssh part and directly execute the program we were trying to execute over ssh?